File: testbed_ruby/test/associations/m_n_to_n_test.rb
require 'test/unit'
require 'src/student_t'
require 'src/mentor_t'
require 'src/program_t'

module CruiseTest
class MNToNTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentT.new(99)
    assert_equal(0,s.number_of_mentors)
  end
  
  def test_CreateMentorWithoutStudent
    m = MentorT.new("blah")
    assert_equal(0,m.number_of_students)
  end
  
  def test_SetStudentsJustRightEnough
    m = MentorT.new("blah")
    
    s = StudentT.new(99)
    s2 = StudentT.new(98)
    s3 = StudentT.new(97)
    
    assert_equal(true,m.set_students([s,s2,s3]))
    
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
  end  
  
  def test_SetStudentsTooMany
    m = MentorT.new("blah")
    
    s = StudentT.new(99)
    s2 = StudentT.new(98)
    s3 = StudentT.new(97)
    s4 = StudentT.new(96)
    
    assert_equal(false,m.set_students([s,s2,s3,s4]))
  end   
  
  def test_AddStudents
    m = MentorT.new("blah")
    
    s = StudentT.new(99)
    s2 = StudentT.new(98)
    s3 = StudentT.new(97)
    s4 = StudentT.new(96)
    s5 = StudentT.new(95)
    s6 = StudentT.new(94)

    assert_equal(true,m.add_student(s))
    assert_equal(true,m.add_student(s2))
    assert_equal(true,m.add_student(s3))
    assert_equal(false,m.add_student(s4))
    assert_equal(false,m.add_student(s5))
    assert_equal(false,m.add_student(s6))
    
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(0,s4.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)
    assert_equal(0,s6.number_of_mentors)
  end   
  
  def test_CannotRemoveStudents
    m = MentorT.new("blah")
    
    s = StudentT.new(99)
    s2 = StudentT.new(98)
    s3 = StudentT.new(97)

    m.set_students([s,s2,s3])
    assert_equal(false,m.remove_student(s3))
  end 
  
  def test_SetStudentsTooManyAndTooFew
    m = MentorT.new("blah")
    
    s = StudentT.new(99)
    s2 = StudentT.new(98)
    s3 = StudentT.new(97)
    s4 = StudentT.new(96)
    s5 = StudentT.new(95)
    
    assert_equal(true,m.set_students([s,s2,s3]))
    
    assert_equal(false,m.set_students([s4,s5]))
    assert_equal(false,m.set_students([s4,s5,s,s2]))
    
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(0,s4.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)
  end   
  
  def test_MentorAlreadyHasEnoughStudents
    m = MentorT.new("blah")
    
    s = StudentT.new(99)
    s2 = StudentT.new(98)
    s3 = StudentT.new(97)
    
    m.set_students([s,s2,s3])
    
    s6 = StudentT.new(94)
    assert_equal(false,m.add_student(s6))
    
    assert_equal(3,m.number_of_students)
    assert_equal(0,s6.number_of_mentors)
  end    

  def test_set_mentors
    m2 = MentorT.new("blah2")
    m3 = MentorT.new("blah3")

    s = StudentT.new(99)
    s2 = StudentT.new(98)
    
    assert_equal(true,s.set_mentors([m2,m3]))
    assert_equal(true,m2.add_student(s2))

    
    assert_equal(2,m2.number_of_students)
    assert_equal(1,m3.number_of_students)
    assert_equal(2,s.number_of_mentors)
  end
 
  def test_is_number_of_students_valid
    m = MentorT.new("blah")
    assert_equal(false,m.is_number_of_students_valid)

    m.add_student(StudentT.new(99))
    assert_equal(false,m.is_number_of_students_valid)

    m.add_student(StudentT.new(98))
    assert_equal(false,m.is_number_of_students_valid)

    m.add_student(StudentT.new(97))
    assert_equal(true,m.is_number_of_students_valid)
  end
  
  def test_getBoundsForStudent
    assert_equal(3,MentorT::minimum_number_of_students)
    assert_equal(3,MentorT::maximum_number_of_students)
  end  
  
  def test_deleteMentorAndStudentHasEnough
    m = MentorT.new("blah")
    m2 = MentorT.new("blah2")
    m3 = MentorT.new("blah3")

    s = StudentT.new(99)
    s2 = StudentT.new(98)
    s3 = StudentT.new(97)
    
    m.set_students([s,s2,s3])
    m2.set_students([s,s2,s3])
    m3.set_students([s,s2,s3])
    
    p = ProgramT.new
    s.set_program(p)
    
    m.delete
    assert_equal(0,m.number_of_students)
    assert_equal(p,s.get_program)
    assert_equal(2,s.number_of_mentors)
  end  

  def test_deleteMentorAndStudentNowHasTooFewMentors
    m = MentorT.new("blah")
    m2 = MentorT.new("blah2")

    s = StudentT.new(99)
    s2 = StudentT.new(98)
    s3 = StudentT.new(97)
    
    m.set_students([s,s2,s3])
    m2.set_students([s,s2,s3])
    
    p = ProgramT.new
    s.set_program(p)
    
    m.delete
    assert_equal(0,m.number_of_students)
    assert_equal(0,m2.number_of_students)
    assert_equal(0,s.number_of_mentors)
  end  

end
end

File: testbed_php/test/associations/MNToNTest.php
<?php

class MNToNTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentT(99);
    $this->assertEqual(0,$s->numberOfMentors());
  }
  

  public function test_CreateMentorWithoutStudent()
  {
    $m = new MentorT("blah");
    $this->assertEqual(0,$m->numberOfStudents());
  }
  

  public function test_SetStudentsJustRightEnough()
  {
    $m = new MentorT("blah");
    
    $s = new StudentT(99);
    $s2 = new StudentT(98);
    $s3 = new StudentT(97);
    
    $this->assertEqual(true,$m->setStudents(array($s,$s2,$s3)));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor_index(0));
    $this->assertEqual($m,$s2->getMentor_index(0));
    $this->assertEqual($m,$s3->getMentor_index(0));
  }  
  

  public function test_SetStudentsTooMany()
  {
    $m = new MentorT("blah");
    
    $s = new StudentT(99);
    $s2 = new StudentT(98);
    $s3 = new StudentT(97);
    $s4 = new StudentT(96);
    
    $this->assertEqual(false,$m->setStudents(array($s,$s2,$s3,$s4)));
  }   
  

  public function test_AddStudents()
  {
    $m = new MentorT("blah");
    
    $s = new StudentT(99);
    $s2 = new StudentT(98);
    $s3 = new StudentT(97);
    $s4 = new StudentT(96);
    $s5 = new StudentT(95);
    $s6 = new StudentT(94);

    $this->assertEqual(true,$m->addStudent($s));
    $this->assertEqual(true,$m->addStudent($s2));
    $this->assertEqual(true,$m->addStudent($s3));
    $this->assertEqual(false,$m->addStudent($s4));
    $this->assertEqual(false,$m->addStudent($s5));
    $this->assertEqual(false,$m->addStudent($s6));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor_index(0));
    $this->assertEqual($m,$s2->getMentor_index(0));
    $this->assertEqual($m,$s3->getMentor_index(0));
    $this->assertEqual(0,$s4->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());
    $this->assertEqual(0,$s6->numberOfMentors());
  }   
  

  public function test_CannotRemoveStudents()
  {
    $m = new MentorT("blah");
    
    $s = new StudentT(99);
    $s2 = new StudentT(98);
    $s3 = new StudentT(97);

    $m->setStudents(array($s,$s2,$s3));
    $this->assertEqual(false,$m->removeStudent($s3));
  } 
  

  public function test_SetStudentsTooManyAndTooFew()
  {
    $m = new MentorT("blah");
    
    $s = new StudentT(99);
    $s2 = new StudentT(98);
    $s3 = new StudentT(97);
    $s4 = new StudentT(96);
    $s5 = new StudentT(95);
    
    $this->assertEqual(true,$m->setStudents(array($s,$s2,$s3)));
    
    $this->assertEqual(false,$m->setStudents(array($s4,$s5)));
    $this->assertEqual(false,$m->setStudents(array($s4,$s5,$s,$s2)));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor_index(0));
    $this->assertEqual($m,$s2->getMentor_index(0));
    $this->assertEqual($m,$s3->getMentor_index(0));
    $this->assertEqual(0,$s4->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());
  }   
  

  public function test_MentorAlreadyHasEnoughStudents()
  {
    $m = new MentorT("blah");
    
    $s = new StudentT(99);
    $s2 = new StudentT(98);
    $s3 = new StudentT(97);
    
    $m->setStudents(array($s,$s2,$s3));
    
    $s6 = new StudentT(94);
    $this->assertEqual(false,$m->addStudent($s6));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(0,$s6->numberOfMentors());
  }    


  public function test_setMentors()
  {
    $m2 = new MentorT("blah2");
    $m3 = new MentorT("blah3");

    $s = new StudentT(99);
    $s2 = new StudentT(98);
    
    $this->assertEqual(true,$s->setMentors(array($m2,$m3)));
    $this->assertEqual(true,$m2->addStudent($s2));

    
    $this->assertEqual(2,$m2->numberOfStudents());
    $this->assertEqual(1,$m3->numberOfStudents());
    $this->assertEqual(2,$s->numberOfMentors());
  } 

  

  public function test_isNumberOfStudentsValid()
  {
    $m = new MentorT("blah");
    $this->assertEqual(false,$m->isNumberOfStudentsValid());

    $m->addStudent(new StudentT(99));
    $this->assertEqual(false,$m->isNumberOfStudentsValid());

    $m->addStudent(new StudentT(98));
    $this->assertEqual(false,$m->isNumberOfStudentsValid());

    $m->addStudent(new StudentT(97));
    $this->assertEqual(true,$m->isNumberOfStudentsValid());
  }
  

  public function test_getBoundsForStudent()
  {
    $this->assertEqual(3,MentorT::minimumNumberOfStudents());
    $this->assertEqual(3,MentorT::maximumNumberOfStudents());
  }  
  

  public function test_deleteMentorAndStudentHasEnough()
  {
    
    $m = new MentorT("blah");
    $m2 = new MentorT("blah2");
    $m3 = new MentorT("blah3");

    $s = new StudentT(99);
    $s2 = new StudentT(98);
    $s3 = new StudentT(97);
    
    $m->setStudents(array($s,$s2,$s3));
    $m2->setStudents(array($s,$s2,$s3));
    $m3->setStudents(array($s,$s2,$s3));
    
    $p = new ProgramT();
    $s->setProgram($p);
    
    $m->delete();
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual($p,$s->getProgram());
    $this->assertEqual(2,$s->numberOfMentors());
  }  


  public function test_deleteMentorAndStudentNowHasTooFewMentors()
  {
    
    $m = new MentorT("blah");
    $m2 = new MentorT("blah2");

    $s = new StudentT(99);
    $s2 = new StudentT(98);
    $s3 = new StudentT(97);
    
    $m->setStudents(array($s,$s2,$s3));
    $m2->setStudents(array($s,$s2,$s3));
    
    $p = new ProgramT();
    $s->setProgram($p);
    
    $m->delete();
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(0,$m2->numberOfStudents());
    $this->assertEqual(0,$s->numberOfMentors());
  }  

}

?>
